interfaces/connection.go

```diff
@@ -177,3 +177,46 @@ func (slot *ConnectedSlot) SetAttr(key string, value interface{}) error {
 func (conn *Connection) Interface() string {
 	return conn.plug.plugInfo.Interface
 }
+
+func copyAttributes(value map[string]interface{}) (map[string]interface{}, error) {
+	cpy, err := copyRecursive(value)
+	if err != nil {
+		return nil, err
+	}
+	return cpy.(map[string]interface{}), err
+}
+
+func copyRecursive(value interface{}) (interface{}, error) {
```

Comment:
Can we have a test with `snap.yaml` that has ~10KB of `{{{{...}}}}` dict to see what happens here? Are we going to exhaust golang stack?

Reply:
we already have recursive code in snap/info_snap_yaml.go so that's probrably a concern there but not here.

About all the copying, my comment was a genunine question, in this full form it feels a bit overkill, is just a bit strange to hide the full map though in a private field but then have an accessor that gives back all of it without even a shallow copy.

I think an important thing I don't understand here is what will typically use Attrs() ?  and what are we going to need to do for dotted/nested attributes with Set* etc?

Reply:
I remembered though we do have a precedent for this kind of copying in accessors, in assertionBase.Header and assertionBase.Headers

Reply:
This API will be used by all interfaces. The point of returning a deep copy of static attributes is valid I think, it's desirable that interfaces code doesn't alter static attributes when it's not supposed to (in *ConnectedPlug/Slot methods).
Regarding dotted/nested attributes, these are handled in snapctl set and at that level we are just internally operating on two dictonaries and not using this API. Not sure if that answers your question?